src/game/boe.dlgutil.cpp

```diff
@@ -194,11 +194,15 @@ void end_shop_mode() {
 	// If it was a random shop, we need to update the stored list of items so that bought items don't reappear
 	if(active_shop.getType() == eShopType::RANDOM) {
 		auto& this_shop = univ.party.magic_store_items[active_shop_num];
-		this_shop.clear();
+		/* FIXME: actually, we expect that the item quantity can be only equal to 1
+		          as in start_shop_mode, store_limited_stock is checked before the
+				  item is loaded from the magic store magic_store_items
+		 */
 		for(int i = 0; i < active_shop.size(); i++) {
 			cShopItem item = active_shop.getItem(i);
-			if(item.type == eShopItemType::TREASURE || item.type == eShopItemType::CLASS || item.type == eShopItemType::OPT_ITEM)
-				this_shop[i] = item.item;
+			if(item.quantity>0 && item.type == eShopItemType::EMPTY &&
+			   this_shop.find(i)!=this_shop.end())
+				this_shop.find(i)->second.variety = eItemType::NO_ITEM;
```

Comment:
Can you explain the reasoning behind this change? You seem to be completely changing the logic.

Reply:
This is the commit: d92113d0869e99aac87660d859e354cdab0c2c2b

I'm ignoring all the porting stuff in that commit for now.

Reply:
There was a problem with the random store; the first time you start shopping, you see the list of items, but if you finish and start shopping again: there are no items.

Basically, modifying this_shop[i] is not valid, because this_shop[i].type will not be valid anymore (i.e. it will not be TREASURE, ...). Here I tried to implement:
`// If it was a random shop, we need to update the stored list of items so that bought items don't reappear
`

When the last item is purchased, takeOne sets item.type to eShopItemType::EMPTY ( and leaves item.quantity=1) => the new condition. And I assume that by setting the variety to NO_ITEM, the item will be hidden.